File: pandemicBot/src/main/java/com/corona/virus/game/PandemicGame.java
package com.corona.virus.game;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

public class PandemicGame {

    public static final String NEW_PLAYER_ADDED = "New player added";

    private GameStatus gameStatus;

    private List<Player> players;

    public PandemicGame(){
        gameStatus = GameStatus.STOPPED;
        players = new ArrayList<>();
    }

    public GameStatus startGame(){
        if (gameStatus == GameStatus.STOPPED){
            gameStatus = GameStatus.STARTED;
        }
        else if (gameStatus == GameStatus.STARTED || gameStatus == GameStatus.IN_PROGRESS){
            gameStatus = GameStatus.IN_PROGRESS;
        }
        else{
            gameStatus = GameStatus.STOPPED;
        }
        return gameStatus;
    }

    public void addPlayerInGame(String playerId){
        players.add(new Player(playerId));
        System.out.println(NEW_PLAYER_ADDED + playerId);
    }

    public boolean isUserCoronned(String authorId) {
        Player currentPlayer = players.stream().filter(player -> player.getId().equals(authorId)).findFirst().get();
        int randomNum = ThreadLocalRandom.current().nextInt(1, 3);
        if (randomNum == 1){
            currentPlayer.setCoronned(true);
            return true;
        }
        else{
            return false;
        }
    }

    public void healPlayers(String authorId) {
        players.forEach((player) -> healPlayer(player, authorId));
    }

    private void healPlayer(Player currentPlayer, String authorId){
        Player healerPlayer = players.stream().filter(player -> player.getId().equals(authorId)).findFirst().get();
        if (currentPlayer.isCoronned()){
            healerPlayer.setScore(healerPlayer.getScore()+1);
            currentPlayer.setCoronned(false);
        }
    }

    public String getGameScores(){
        StringBuilder scoresStringBuilder = new StringBuilder();
        players.forEach((player) -> scoresStringBuilder.append(player.toString()));
        return scoresStringBuilder.toString();
    }

    public GameStatus getGameStatus() {
        return gameStatus;
    }

    public void setGameStatus(GameStatus gameStatus) {
        this.gameStatus = gameStatus;
    }
}

File: pandemicBot/src/main/kotlin/com/corona/virus/game/PandemicGame.kt
package com.corona.virus.game

import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import org.springframework.stereotype.Component
import java.time.OffsetDateTime
import java.util.concurrent.ThreadLocalRandom
import kotlin.math.absoluteValue

@Component
class PandemicGame {

    var gameStatus = GameStatus.STOPPED
    private val players = ArrayList<Player>()
    var isHealAction = false
    var healMessagesHistory = LinkedHashMap<String, Pair<OffsetDateTime, Boolean>>()

    fun startGame(): GameStatus {
        gameStatus = if (gameStatus === GameStatus.STOPPED) {
            GameStatus.STARTED
        } else if (gameStatus === GameStatus.STARTED || gameStatus === GameStatus.IN_PROGRESS) {
            GameStatus.IN_PROGRESS
        } else {
            GameStatus.STOPPED
        }
        return gameStatus
    }

    fun addPlayerInGame(playerId: String, playerName: String, playerScore: Int) {
        var player = Player(playerId)
        player.playerName = playerName
        player.score = playerScore
        players.add(player)
    }

    fun isUserCoronned(authorId: String): Boolean {
        val currentPlayer = players.stream().filter { player: Player -> player.id == authorId }.findFirst().get()
        val randomNum = ThreadLocalRandom.current().nextInt(1, 10)
        return if (randomNum == 1 && !currentPlayer.isCoronned) {
            currentPlayer.isCoronned = true
            true
        } else {
            false
        }
    }

    fun healPlayers(authorId: String) {
        players.forEach{ healPlayer(it, authorId) }
    }

    private fun healPlayer(currentPlayer: Player, authorId: String) {
        val healerPlayer = players.stream().filter { it.id == authorId }.findFirst().get()
        if (currentPlayer.isCoronned) {
            healerPlayer.score++
            currentPlayer.isCoronned = false
            isHealAction = true
        }
    }

    fun getPlayers() =  players

    fun isChehCommandEffective(event: MessageReceivedEvent, authorId: String): Boolean {
        val healMessageHistoryEntries = healMessagesHistory.entries
        val lastHealAuthor = healMessageHistoryEntries.last().key
        val beforeLastHealAuthor = healMessageHistoryEntries.elementAt(healMessageHistoryEntries.size-2).key
        val lastHealMessageEffective = healMessageHistoryEntries.last().value.second
        val beforeLastHealMessageEffective = healMessageHistoryEntries.elementAt(healMessageHistoryEntries.size-2).value.second
        return (lastHealAuthor != beforeLastHealAuthor) &&
            ( (!lastHealMessageEffective && beforeLastHealMessageEffective) ||
            (lastHealMessageEffective && !beforeLastHealMessageEffective) )
    }

    fun processCheh(authorId: String) {
        val cheherPlayer = players.stream().filter { it.id == authorId }.findFirst().get()
        cheherPlayer.score+=2
    }


}